weaviate/collections/classes/batch.py

```diff
@@ -165,26 +175,56 @@ class BatchObjectReturn:
             A boolean indicating whether or not any of the objects in the batch failed to be inserted. If this is `True`, then the `errors` dictionary will contain at least one entry.
     """
 
-    all_responses: List[Union[uuid_package.UUID, ErrorObject]]
+    _all_responses: List[Union[uuid_package.UUID, ErrorObject]]
     elapsed_seconds: float
+    """The time taken to perform the batch operation."""
     errors: Dict[int, ErrorObject]
+    """A dictionary of all the failed responses from the batch operation. The keys are the indices of the objects in the overall batch, and the values are the `Error` objects."""
     uuids: Dict[int, uuid_package.UUID]
+    """A dictionary of all the successful responses from the batch operation. The keys are the indices of the objects in the overall batch, and the values are the `uuid_package.UUID` objects."""
     has_errors: bool = False
+    """A boolean indicating whether or not any of the objects in the batch failed to be inserted. If this is `True`, then the `errors` dictionary will contain at least one entry."""
+
+    @property
+    def all_responses(self) -> List[Union[uuid_package.UUID, ErrorObject]]:
+        """@deprecated: A list of all the responses from the batch operation. Each response is either a `uuid_package.UUID` object or an `Error` object.
+
+        WARNING: Now that only the last `MAX_STORED_RESULTS` uuids are stored in the `uuids` dictionary, the keys of the `errors` and `uuids` dictionaries can no longer be used to map the indices in the `all_responses` field!
+        """
+        _Warnings.batch_results_objects_all_responses_attribute()
+        return self._all_responses
 
     def __add__(self, other: "BatchObjectReturn") -> "BatchObjectReturn":
-        self.all_responses += other.all_responses
+        self._all_responses += other._all_responses
 
         prev_max = max(self.errors.keys()) if len(self.errors) > 0 else -1
         for k1, v1 in other.errors.items():
-            self.errors[prev_max + k1] = v1
+            self.errors[prev_max + k1 + 1] = v1
 
         prev_max = max(self.uuids.keys()) if len(self.uuids) > 0 else -1
-        for k1, v2 in other.uuids.items():
-            self.uuids[prev_max + k1] = v2
+        for k2, v2 in other.uuids.items():
+            self.uuids[prev_max + k2 + 1] = v2
 
         self.has_errors = self.has_errors or other.has_errors
+
+        if len(self.uuids.keys()) > MAX_STORED_RESULTS:
+            new_max = max(self.uuids.keys())
+            new_uuid_indices = range(new_max - MAX_STORED_RESULTS + 1, new_max + 1)
+            self.uuids = {index: self.uuids[index] for index in new_uuid_indices}
+            self._all_responses = [
+                self.__error_or_uuid(idx) for idx in merge(self.errors.keys(), self.uuids.keys())
+            ]
```

Comment:
This means we have to rebuild the dict/list every time we add more entries above MAX_STORED_RESULTS.

Dicts keep their order and this is guranteed as fasr as I know (https://mail.python.org/pipermail/python-dev/2017-December/151283.html) and we could use that to drop on one side and append on the other?

Reply:
hmm, we are still copying `self.uuids.keys()` and recreating self._all_responses.

For uuids: We know which keys will fall out (the lowest one until lowest+number_new_entries)
For all_responses: Can't we just do self._all_responses[number_that_needs_to_be_dropped:] and then append the new things afterwards?

Reply:
For `self.uuids.keys()`, it has to be copied before we loop through and delete keys otherwise the size of the dictionary changes in each loop as we delete

For `self._all_responses`, not necessarily since the uuids and errors could be interspersed between each other so we need to ensure that the uuids are the ones being dropped but the errors are maintained. If we did `self._all_responses[n_dropped:]` we may accidentally remove errors, no?

Or is your idea that we just keep `MAX_NUMBER` in `all_responses` but keep all errors in `self.errors` with their original index as the key? I like this idea but hadn't implemented it this way!